skyportal/plot.py

```diff
@@ -1398,397 +1400,423 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
     if len(spectra) == 0:
         return None, None, None
 
-    rainbow = cm.get_cmap('rainbow', len(spectra))
-    palette = list(map(rgb2hex, rainbow(range(len(spectra)))))
-    color_map = dict(zip([s.id for s in spectra], palette))
-
-    data = []
-    for i, s in enumerate(spectra):
-
-        # normalize spectra to a median flux of 1 for easy comparison
-        normfac = np.nanmedian(np.abs(s.fluxes))
-        normfac = normfac if normfac != 0.0 else 1e-20
-        altdata = json.dumps(s.altdata) if s.altdata is not None else ""
-        df = pd.DataFrame(
-            {
-                'wavelength': s.wavelengths,
-                'flux': s.fluxes / normfac,
-                'id': s.id,
-                'telescope': s.instrument.telescope.name,
-                'instrument': s.instrument.name,
-                'date_observed': s.observed_at.isoformat(sep=' ', timespec='seconds'),
-                'pi': (
-                    s.assignment.run.pi
-                    if s.assignment is not None
-                    else (
-                        s.followup_request.allocation.pi
-                        if s.followup_request is not None
-                        else ""
-                    )
-                ),
-                'origin': s.origin,
-                'altdata': altdata[:20] + "..." if len(altdata) > 20 else altdata,
-            }
+    layouts = []
+    spectrum_types = []
+    for spec_type in ALLOWED_SPECTRUM_TYPES:
+        spectra_of_same_type = [s for s in spectra if s.type == spec_type]
```

Comment:
It looks like you want to build a dictionary of spectrum_type -> spectra, and then iterate over that.

Reply:
Is there a clever way to do that other than a list comprehension expression that sorts each spectrum into the right label?

Reply:
```
In [1]: spectra = [{'type': 'one', 'data': 1}, {'type': 'two', 'data': 2}]

In [2]: import collections

In [3]: spectra_by_type = collections.defaultdict(list)

In [4]: for s in spectra:
   ...:     spectra_by_type[s['type']].append(s)

```